File: practice/hosh.py
# a unix shell by Hobson
# based on a C program in Advanced Programming in the UNIX(R) Environment, Third Edition
# section 1.6

import sys  # argv, stdin, stdout, stderr
import os   # fork


def main(argv):
    sys.stdout.write("%% ")
    buf = sys.stdin.readline()
    while (buf and buf is not '\n'):
        pid = 0
        try:
            pid = os.fork()
        except:
            print('fork error')
        if pid == 0:  # child
            bufs = buf.split(' ')
            if len(bufs) < 2:
                bufs += ['']
            retval = os.execlp(*bufs)
            if retval >= 0:
                return 127
            sys.stderr.write("couldn't execute: %s" % buf)
            return 0
        elif pid:
            try:
                status = os.waitpid(pid, 0)
            except OSError as e:
                sys.stderr.write("I/O error({0}): {1}".format(e.errno, e.strerror))
                status = (pid, -1)
            if status and len(status) == 2 and status[1] < 0:
                sys.stderr.write("waitpid error")
        sys.stdout.write("%% ")
        buf = sys.stdin.readline()
    return 0


if __name__ == '__main__':
    sys.exit(main(sys.argv))

File: practice/hosh.cpp
// programming challenge 110101
// http://www.programming-challenges.com/pg.php?page=studenthome
// 3n + 1 problem


#include <stdio.h>
#include <sys/types.h>      /* some systems still require this */
#include <sys/stat.h>
#include <sys/termios.h>    /* for winsize */

#if defined(MACOS) || !defined(TIOCGWINSZ)
#include <sys/ioctl.h>
#endif

#include <stdio.h>      /* for convenience */
#include <stdlib.h>     /* for convenience */
#include <stddef.h>     /* for offsetof */
#include <string.h>     /* for convenience */
#include <unistd.h>     /* for convenience */
#include <signal.h>     /* for SIG_ERR */


#define MAXLINE 4096            /* max line length */
#include <sys/wait.h>

int
main(void)
{
    char    buf[MAXLINE];   /* from apue.h */
    pid_t   pid;
    int     status;

    printf("%% ");  /* print prompt (printf requires %% to print %) */
    while (fgets(buf, MAXLINE, stdin) != NULL) {
        if (buf[strlen(buf) - 1] == '\n')
            buf[strlen(buf) - 1] = 0; /* replace newline with null */

        if ((pid= fork()) < 0) {
            err_sys("fork error");
        } else if (pid== 0) {      /* child */
            execlp(buf, buf, (char *)0);
            err_ret("couldn't execute: %s", buf);
            exit(127);
        }

        /* parent */
        if ((pid= waitpid(pid, &status, 0)) < 0)
            err_sys("waitpid error");
        printf("%% ");
    }
    exit(0);

